sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/AvroUtils.java

```diff
@@ -428,6 +432,35 @@ public static GenericRecord toGenericRecord(
     return new GenericRecordToRowFn(schema);
   }
 
+  public static Row avroBytesToRow(byte[] bytes, Schema schema) {
+    try {
+      org.apache.avro.Schema avroSchema = AvroUtils.toAvroSchema(schema);
+      AvroCoder<GenericRecord> coder = AvroCoder.of(avroSchema);
```

Comment:
Both of these methods will repeat the schema conversion for every Row. Instead both of these methods should convert the schema once and re-use `avroSchema` and `coder` for every instance of `bytes` or `row` that comes in. There are a few ways to do this, I don't really have a preference for which:
- Return a lambda with `avroSchema` and `coder` in it's closure, like 
```java
return (bytes) -> {
  ByteArrayInputStream inputStream = new ByteArrayInputStream(bytes);
  GenericRecord record = coder.decode(inputStream);
  AvroUtils.toBeamRowStruct(record, schema);
}
```
- Create an AvroBytesToRow class with a process method that re-uses avroSchema/coder
- Similarly, create an AvroBytesToRow DoFn with a process method that re-uses avroSchema/coder

Reply:
The last is probably the most natural for Beam

Reply:
I had a quite hard time as AvroSchema is not serializable (and the tests were failing due to DoFn serialization exception), but I've learned the use case of transient keyword :) I've chosen the lambda way but had to define a class implementing SerializableFunction to be able to use transient for avroSchema. Done.

Reply:
Looks good, thanks for working through it :)